wallet-api/src/Wallet/Emulator/Http.hs

```diff
@@ -0,0 +1,164 @@
+import           Servant.API                ((:<|>) ((:<|>)), (:>), Capture, Get, JSON, NoContent (NoContent), Post,
+                                             ReqBody)
+import qualified Wallet.API                 as WAPI
+import           Wallet.Emulator.Types      (EmulatedWalletApi, Wallet (Wallet), WalletState, emptyWalletState,
+                                             getWallet, runEmulatedWalletApi)
+import           Wallet.UTXO                (Tx, TxIn', TxOut', Value)
+
+type WalletAPI
+   = "wallets" :> Get '[ JSON] [Wallet]
+     :<|> "wallets" :> Capture "walletid" Int :> Get '[ JSON] Wallet
+     :<|> "wallets" :> ReqBody '[ JSON] Wallet :> Post '[ JSON] NoContent
+     :<|> "wallets" :> Capture "walletid" Int :> "payments" :> ReqBody '[ JSON] Value :> Post '[ JSON] ( Set TxIn'
+                                                                                                       , TxOut')
+     :<|> "wallets" :> Capture "walletid" Int :> "pay-to-public-key" :> ReqBody '[ JSON] Value :> Post '[ JSON] TxOut'
+     :<|> "wallets" :> Capture "walletid" Int :> "transactions" :> ReqBody '[ JSON] Tx :> Post '[ JSON] ()
+     :<|> "wallets" :> "transactions" :> Get '[ JSON] [Tx]
+
+newtype State = State
+  { getWallets :: TVar (HashMap Int WalletState)
+  }
+
+data AppError
+  = WalletAPIError WAPI.WalletAPIError
+  | HttpError ServantErr
+
+wallets ::
+     (MonadError ServantErr m, MonadReader State m, MonadIO m) => m [Wallet]
+wallets = do
+  var <- asks getWallets
+  ws <- liftIO $ readTVarIO var
+  pure $ Wallet <$> HM.keys ws
+
+fetchWallet ::
+     (MonadError ServantErr m, MonadReader State m, MonadIO m)
+  => Int
+  -> m Wallet
+fetchWallet walletId = do
+  var <- asks getWallets
+  ws <- liftIO $ readTVarIO var
+  if HM.member walletId ws
+    then pure (Wallet walletId)
+    else throwError err404
+
+createWallet :: (MonadReader State m, MonadIO m) => Wallet -> m NoContent
+createWallet wallet = do
+  var <- asks getWallets
+  let walletState = emptyWalletState wallet
+  liftIO . atomically $
+    modifyTVar var (HM.insert (getWallet wallet) walletState)
+  pure NoContent
+
+createPaymentWithChange ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m (Set.Set TxIn', TxOut')
+createPaymentWithChange = runWalletApiAction WAPI.createPaymentWithChange
+
+payToPublicKey ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Value
+  -> m TxOut'
+payToPublicKey = runWalletApiAction WAPI.payToPublicKey
+
+submitTxn ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => Int
+  -> Tx
+  -> m ()
+submitTxn = runWalletApiAction WAPI.submitTxn
+
+runWalletApiAction ::
+     (MonadReader State m, MonadIO m, MonadError ServantErr m)
+  => (a -> EmulatedWalletApi b)
+  -> Int
+  -> a
+  -> m b
+runWalletApiAction action walletId a = do
+  var <- asks getWallets
+  finalResult <-
+    liftIO . atomically $ do
+      states <- readTVar var
```

Comment:
It feels to me like you might want to spin this do block out into its own top-level function. Otherwise there's just a slightly odd separation between the `pure . Left . HttpError $ err404` and the `Let (HttpError e) -> throwError e`.

I see why they have to be separate (because of the throw), but given you've gone to the trouble of separating them already, and things are getting a little unwieldy, I think going for the full separation might clarify each step. 2c.

Reply:
done